doc/subsystems/networking/bsd-sockets.rst

```diff
@@ -0,0 +1,65 @@
+##########################
+
+Zephyr offers an implementation of a subset of BSD Sockets API (a part
+of POSIX standard). This allows to reuse existing programming experience
+and port existing simple networking applications to Zephyr.
+
+Following are the requirements and concepts which governed BSD Sockets
+compatible API implementation for Zephyr:
+
+* Should have minimal overhead, similar to the requirement for other
+  Zephyr subsystems.
+* Should be implemented on top of
+  :ref:`native networking API <networking_api_usage>` to keep modular
+  design.
+* Should be namespaced by default, to avoid name conflicts with well-known
+  names like ``close()`` which may be part of libc or other POSIX
+  compatibility libraries. If enabled, should also expose native POSIX
+  names.
+
+BSD Sockets compatible API is enabled using :option:`CONFIG_NET_SOCKETS`
+config option and implements following operations: ``socket()``,
+``close()``, ``recv()``, ``send()``, ``connect()``, ``bind()``,
+``listen()``, ``fcntl()`` (to set non-blocking mode), ``poll()``.
+
+Based on the namespacing requirements above, these operations are by
+default exposed as functions with ``zsock_`` prefix, e.g.
+:c:func:`zsock_socket()` and :c:func:`zsock_close()`. If config option
+:option:`CONFIG_NET_SOCKETS_POSIX_NAMES` is defined, all the functions
+will be also exposed as aliases without the prefix. This includes the
+functions like ``close()`` and ``fcntl()`` (which may conflict with
+functions in libc or other libraries (e.g. with filesystem libraries)).
+
+Native BSD Sockets API uses file descriptors to represent sockets,
+which are known to be integers consecutively assigned from zero (and
+thus usually small). Internally, there is usually a table mapping
+file descriptors to internal object pointers. For memory efficiency
+reasons, Zephyr BSD Sockets compatible API is devoid of such table.
+Instead, `net_context` pointers, cast to an int, are used to represent
+sockets. Thus, socket identifiers aren't really small integers. One
+entailment of this is that ``select()`` operation is not available, as
+it depends on "small int" property of the file descriptors. Instead of
+it, ``poll()`` operation is available which is in general more efficient.
+
+BSD Sockets API (and POSIX API in general) also treat negative file
+descriptors values in a special way (such values usually mean an
+error). As Zephyr API uses pointer value case to an int, it means
+that the pointer should not have the highest bit set, in other words,
+pointers should not refer to 2nd (highest) part of the address space.
+For many CPU architectures and SoCs Zephyr supports, user RAM is
+located in the lower half, so the above condition is satisfied. If
+you face an issue with some SoC, please report it to the Zephyr bug
+tracker or mailing list. The decision to use pointers to represent
+sockets can be reworked at later time.
+
+The final entailment of the design requirements above is that Zephyr
```

Comment:
The Zephyr API aggressively employs the short-read/short-write property of the POSIX API